### PR TITLE
[no sq] Add binary glTF (.glb) support + other glTF improvements

### PR DESCRIPTION
**Pull request by appgurueu**
Adds glb support for slightly smaller models which don't need their buffers to be base64-decoded. (The buffers are still copied even though for our purposes this isn't necessary, because we don't want to tie the lifetime of tiniergltf structs to that of the buffer they were read from).

## To do

This PR is Ready for Review.

## How to test

* Compare glb vs glTF blender cube in-game
* Verify that unit tests pass
* Throw some more binary glTFs at it

---

cc @sfan5 @Desour since you expressed interest in this

### CODE REVIEW

**Comment by sfan5** [comment]
Should I review this or the other PR first?

**Comment by appgurueu** [comment]
> Should I review this or the other PR first?

Whichever you prefer. Personally I think animated gltf is higher priority than glb support, but then again this PR is a bit simpler. They don't really conflict much if at all.

File: doc/lua_api.md [hunk 1]
```diff
@@ -274,7 +274,7 @@ Accepted formats are:
 
     images: .png, .jpg, .tga, (deprecated:) .bmp
     sounds: .ogg vorbis
-    models: .x, .b3d, .obj, .gltf (Minetest 5.10 or newer)
+    models: .x, .b3d, .obj, .gltf, .glb (Minetest 5.10 or newer)
```
Comment (grorp) [inline doc/lua_api.md:277]: it's ambigous what model formats `(Minetest 5.10 or newer)` refers to, putting it before with a colon like above would be better

File: lib/tiniergltf/tiniergltf.hpp [hunk 2]
```diff
@@ -1354,4 +1370,123 @@ struct GlTF {
...
+	}
+};
+
+static Json::Value readJson(Span<const char> span) {
+	Json::CharReaderBuilder builder;
+	const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
+	Json::Value json;
+	JSONCPP_STRING err;
+	if (!reader->parse(span.ptr, span.end(), &json, &err))
+		throw std::runtime_error("invalid JSON");
```
Comment (grorp) [inline lib/tiniergltf/tiniergltf.hpp:1396]: shouldn't the error message be forwarded? currently, you get:

```
ERROR[Main]: Irrlicht: glTF loader: invalid JSON
```

```suggestion
		throw std::runtime_error(std::string("invalid JSON: ").append(err));
```

would give

```
ERROR[Main]: Irrlicht: glTF loader: invalid JSON: * Line 2, Column 1
  Missing '}' or object member name

```

File: lib/tiniergltf/tiniergltf.hpp [hunk 3]
```diff
@@ -1354,4 +1370,123 @@ struct GlTF {
...
+			if (span.len < chunkLen)
+				throw std::runtime_error("premature EOF");
+			advance(chunkLen);
+			return {chunkType, {chunkPtr, chunkLen}};
+		}
+	};
+
+	constexpr uint32_t MAGIC_glTF = 0x46546C67;
+	constexpr uint32_t MAGIC_JSON = 0x4E4F534A;
+	constexpr uint32_t MAGIC_BIN = 0x004E4942;
```
Comment (grorp) [inline lib/tiniergltf/tiniergltf.hpp:1444]: I would have spent less time looking at this code if it wasn't inside the function

File: lib/tiniergltf/tiniergltf.hpp [hunk 4]
```diff
@@ -1093,6 +1104,12 @@ struct Texture {
 };
 template<> Texture as(const Json::Value &o) { return o; }
 
+using UriResolver = std::function<std::string(const std::string &uri)>;
+static std::string uriError(const std::string &uri) {
```
Comment (grorp) [inline lib/tiniergltf/tiniergltf.hpp:1108]: compiler warning:

```
In file included from /mt/irr/src/CSceneManager.cpp:23:
In file included from /mt/irr/src/CGLTFMeshFileLoader.h:13:
/mt/lib/tiniergltf/tiniergltf.hpp:1108:20: warning: 'static' function 'uriError' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
 1108 | static std::string uriError(const std::string &uri) {
      |                    ^~~~~~~~
1 warning generated.
```

**Review by grorp — COMMENTED** [review]
Read some of the gltf specification, stuff checks out. Just some minor comments.

File: lib/tiniergltf/tiniergltf.hpp (hunk 2 above)
Comment (appgurueu) [reply lib/tiniergltf/tiniergltf.hpp]: sure, that's better

File: doc/lua_api.md (hunk 1 above)
Comment (appgurueu) [reply doc/lua_api.md]: of course, done

File: lib/tiniergltf/tiniergltf.hpp (hunk 3 above)
Comment (appgurueu) [reply lib/tiniergltf/tiniergltf.hpp]: Well, I do really want to namespace these definitions to within the function; I don't want to expose them. The function scope fits this perfectly.

**Review by grorp — APPROVED** [review]
(only tested with the included model)

**Review by Desour — COMMENTED** [review]
Otherwise looks good from what I can tell. :+1: 

Some things untouched by this PR that I noticed while reading through code:
* The `CGLTFMeshFileLoader::MeshExtractor::MeshExtractor(...)` constructor copies the gltf model.
* It's not checked whether the required extensions of a gltf file are fullfilled. (I.e. if it's empty.)

(For actually smaller files, we still need compression. What do you think about supporting [draco](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_draco_mesh_compression/README.md), btw.?)

**Review by Desour — APPROVED** [review]

**Comment by appgurueu** [comment]
> Some things untouched by this PR that I noticed while reading through code: [...]

Thanks for catching these. I'm lazy so I just added another tiny commit fixing these to this PR and marked it no squash :^)

> (For actually smaller files, we still need compression. What do you think about supporting [draco](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_draco_mesh_compression/README.md), btw.?)

Sure, looks simple enough and [worthwhile](https://www.ultraengine.com/community/blogs/entry/2727-analysis-of-google-draco/). I can work on that sometime after #14685 is merged.

**Comment by appgurueu** [comment]
Threw [another glb test model](https://sketchfab.com/3d-models/pumpkin-9b1dd1cbe57e4bf7a8a4b1386250f2d4#download) off the webs at it. Failed because the gltf code doesn't support texture wrapping yet, so I implemented that, resulting in https://github.com/minetest/minetest/pull/15121/commits/643b79e35446c5d0c64c81c38d10268b0a47398d. Now the pumpkin works fine.

File: irr/include/vector2d.h [hunk 5]
```diff
@@ -34,6 +35,9 @@ class vector2d
 	constexpr vector2d(const dimension2d<T> &other) :
 			X(other.Width), Y(other.Height) {}
 
+	constexpr vector2d(const std::array<T, 2> &arr) :
```
Comment (Desour) [inline irr/include/vector2d.h:38]: Maybe mark it explicit.

Comment (appgurueu) [reply irr/include/vector2d.h:38]: Personally I consider this one of a few places where an implicit conversion wouldn't surprise me, but sure

**Review by Desour — APPROVED** [review]

**Pull request merged**